File: Recursion/GFG/Medium/ReverseStack/ReverseStack.cpp
// https://www.geeksforgeeks.org/problems/sort-a-stack/
#include<bits/stdc++.h>
using namespace std;

void insertReversed(stack<int> &st, int x)
{
    if (st.empty())
    {
        st.push(x);
    }
    int y = st.top();
    st.pop();
    insertReversed(st, x);
    st.push(y);
}
void Reverse(stack<int> &st)
{
    if (!st.empty())
    {
        int x = st.top();
        st.pop();
        Reverse(st);
        stack<int> temp;
        while (!st.empty())
        {
            temp.push(st.top());
            st.pop();
        }
        st.push(x);
        while (!temp.empty())
        {
            st.push(temp.top());
            temp.pop();
        }
    }
}
int main(void) {
    stack<int> st;
    st.push(1);
    st.push(2);
    st.push(3);
    st.push(4);
    st.push(5);
    st.push(6);
    Reverse(st);
    while(!st.empty()) {
        cout << st.top() << endl;
        st.pop();
    }
    return 0;
}

File: Recursion/GFG/Medium/ReverseStack/ReverseStack.java
// https://www.geeksforgeeks.org/problems/sort-a-stack/
import java.util.Stack;

public class ReverseStack {
    static void endInsert(Stack<Integer> s, int x) {
        if(s.isEmpty()) {
            s.push(x);
            return;
        }
        int y = s.pop();
        endInsert(s, x);
        s.push(y);
    }
    static void reverse(Stack<Integer> s) {
        if(s.isEmpty()) {
            return;
        }
        int x = s.pop();
        reverse(s);
        endInsert(s, x);
    }
    public static void main(String[] args) {
        Stack<Integer> st = new Stack<>();
        st.push(1);
        st.push(2);
        st.push(3);
        st.push(4);
        st.push(5);
        st.push(6);
        System.out.println(st.toString());
        reverse(st);
        System.out.println(st.toString());
    }
}
